ci/ci-test-tasks/main.yml

```diff
@@ -6,78 +6,150 @@ parameters:
 - name: BuildSourceVersion
   displayName: The latest version control change of the triggering repo that is included in this build.
   type: string
+- name: PushNugetPackage
+  displayName: A flag for uploading nuget package
+  type: boolean
+- name: PublishHotfixArtifact
+  displayName: A flag for publishing hotfix artifacts
+  type: boolean
 
 trigger: none
 pr: none
 
 stages:
-- stage: Build
+- stage: ProcessAllTasks
   condition: eq(variables['build.reason'], 'Manual')
   jobs:
-    - ${{ each task in split(parameters.tasks, ',') }}:
-      - job: ${{ task }}_build_task
-        displayName: Build Task (${{ task }})
-        pool:
-          vmImage: windows-2022
-        steps:
-        - template: ./build-task.yml
-          parameters:
-            task: ${{ task }}
-            push: false
-            publishArtifact: true
-            patchVersion: true
-            version: '0.999.0'
-            BuildSourceVersion: ${{ parameters.BuildSourceVersion }}
+    - ${{ if not(parameters.PublishHotfixArtifact) }}:
+      - ${{ each task in split(parameters.tasks, ',') }}:
+        - job:  ${{ task }}_build_and_test
+          displayName: Process ${{ task }}
+          pool:
+            vmImage: windows-2022
+          steps:
+          # Start collect diagnostics
+          - powershell: ./ci/start-collect-diagnostics.ps1
+            displayName: Start collect diagnostics
 
-- stage: PreDelete
-  jobs:
-    - ${{ each task in split(parameters.tasks, ',') }}:    
-      - job: ${{ task }}_pre_delete
-        displayName: PreDelete (${{ task }})
-        pool:
-          vmImage: windows-2022
-        steps:
-        - template: ./delete-task.yml
-          parameters:
-            task: ${{ task }}
-            BuildSourceVersion: ${{ parameters.BuildSourceVersion }}
+          - template: ./build-init.yml
+            parameters:
+              BuildSourceVersion: ${{ parameters.BuildSourceVersion }}
 
-- stage: Upload
-  jobs:
-    - ${{ each task in split(parameters.tasks, ',') }}:
-      - job: ${{ task }}_upload_task
-        displayName: Upload (${{ task }})
-        pool:
-          vmImage: windows-2022
-        steps:
-        - template: ./upload-task.yml
-          parameters:
-            task: ${{ task }}
-            BuildSourceVersion: ${{ parameters.BuildSourceVersion }}
+          - template: ./build-task.yml
+            parameters:
+              task: ${{ task }}
+              patchVersion: true
+              version: '0.999.0'
 
-- stage: Test
-  jobs:
-    - ${{ each task in split(parameters.tasks, ',') }}:
-      - job: ${{ task }}_test_task
-        displayName: Test (${{ task }})
+          # Stop collect diagnostics
+          - powershell: ./ci/stop-collect-diagnostics.ps1
+            displayName: Stop collect diagnostics
+            condition: always()
+
+          - script: |
```

Comment:
Here for each task we install tfx-cli and login? I am afraid it will take way to long on PR's with a number of tasks. Some time we have PRs with tens of tasks changed.

Reply:
Yes, but each task is being executed on a different build agent since it's a separate job, so each build machine has to have its own version of tfx installed to proceed.